src/Compilers/CSharp/Portable/Symbols/TypeParameterSymbol.cs

```diff
@@ -452,35 +452,47 @@ internal bool IsReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAn
             bool? result = false;
             foreach (TypeSymbolWithAnnotations constraintType in constraintTypes)
             {
-                if (constraintType.IsAnnotated)
+                bool? fromType = IsNotNullableIfReferenceTypeFromConstraintType(constraintType, inProgress);
```

Comment:
>IsNotNullableIfReferenceTypeFromConstraintType [](start = 33, length = 46)

Nice. I'd been thinking about such refactoring as well.
I think a similar simplification (making the top-level logic more apparent) could be applied to `GetIsNotNullableIfReferenceType` below.

```C#
        internal bool? GetIsNotNullableIfReferenceType(ConsList<TypeParameterSymbol> inProgress)
        {
            if (inProgress.ContainsReference(this))
            {
                return false;
            }
 
            bool sawOblivious = false; // or could be `bool? result` as you did
            if (this.HasReferenceTypeConstraint)
            {
                switch (ReferenceTypeConstraintIsNullable)
                {
                    case false:
                        return true;
                    case null:
                        sawOblivious = true;
                        break;
                }
            }
 
            ImmutableArray<TypeSymbolWithAnnotations> constraintTypes = this.GetConstraintTypesNoUseSiteDiagnostics(inProgress, early: true);
            if (!constraintTypes.IsEmpty)
            {
                switch (IsNotNullableIfReferenceTypeFromConstraintTypes(constraintTypes, inProgress))
                {
                    case true:
                        return true;
                    case null:
                        sawOblivious = true;
                        break;
                }
            }
 
            return sawOblivious ? (bool?)null : false;
        }
```